SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp

```diff
@@ -394,78 +375,68 @@ bool USpatialLatencyTracer::BeginLatencyTrace_Internal(const FString& TraceDesc,
+		return false;
 	}
 
+	// We do minimal internal tracking for native rpcs/properties
+	const bool bInternalTracking = GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking() || Type == ETraceType::Tagged;
+
 	FScopeLock Lock(&Mutex);
-		
-	TraceSpan* ActiveTrace = GetActiveTraceOrReadPayload(LatencyPayload);
-	if (ActiveTrace == nullptr)
+
+	OutLatencyPayload = LatencyPayload;
+	if (OutLatencyPayload.Key == InvalidTraceKey)
 	{
-		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to continue (%s)"), *WorkerId, *TraceDesc);
-		return false;
+		ResolveKeyInLatencyPayload(OutLatencyPayload);
 	}
 
-	TraceKey Key = CreateNewTraceEntry(Actor, Target, Type);
-	if (Key == InvalidTraceKey)
+	const TraceKey Key = OutLatencyPayload.Key;
+	const TraceSpan* ActiveTrace = TraceMap.Find(Key);
+	if (ActiveTrace == nullptr)
 	{
-		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+		UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : No active trace to continue (%s)"), *WorkerId, *TraceDesc);
 		return false;
 	}
 
-	WriteKeyFrameToTrace(ActiveTrace, TCHAR_TO_UTF8(*TraceDesc));
-	WriteKeyFrameToTrace(ActiveTrace, FString::Printf(TEXT("Continue trace %s : %s"), *UEnum::GetValueAsString(Type), *Target));
-
-	// For non-spatial tracing
-	const improbable::trace::SpanContext& TraceContext = ActiveTrace->context();
-
+	if (bInternalTracking)
 	{
-		TArray<uint8> TraceBytes = TArray<uint8_t>((const uint8_t*)&TraceContext.trace_id()[0], sizeof(improbable::trace::TraceId));
-		TArray<uint8> SpanBytes = TArray<uint8_t>((const uint8_t*)&TraceContext.span_id()[0], sizeof(improbable::trace::SpanId));
-		OutLatencyPayloadContinue = FSpatialLatencyPayload(MoveTemp(TraceBytes), MoveTemp(SpanBytes));
+		if (!AddTrackingInfo(Actor, Target, Type, Key))
+		{
+			UE_LOG(LogSpatialLatencyTracing, Warning, TEXT("(%s) : Failed to create Actor/Func trace (%s)"), *WorkerId, *TraceDesc);
+			return false;
+		}
 	}
 
-	// Move the active trace to a new tracked trace
-	TraceSpan TempSpan(MoveTemp(*ActiveTrace));
-	TraceMap.Add(Key, MoveTemp(TempSpan));
-	TraceMap.Remove(ActiveTraceKey);
-	ActiveTraceKey = InvalidTraceKey;
-	PayloadToTraceKeys.Remove(LatencyPayload);
-	PayloadToTraceKeys.Add(MakeTuple(OutLatencyPayloadContinue, Key)); // Add continued payload to tracking map. 
+	WriteKeyFrameToTrace(ActiveTrace, FString::Printf(TEXT("Continue [%s] %s - %s"), *TraceDesc, *UEnum::GetValueAsString(Type), *Target));
 
-	if (!GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
+	// If we're not doing any further tracking, end the trace
+	if (!bInternalTracking)
 	{
-		// We can't do any deeper tracing in the stack here so terminate these traces here
-		if (Type == ETraceType::RPC || Type == ETraceType::Property)
-		{
-			EndLatencyTrace(Key, TEXT("End of native tracing"));
-		}
-
-		ClearTrackingInformation();
+		WriteAndEndTraceIfRemote(Key, TEXT("Native - End of Tracking"));
```

Comment:
Can this Key be the one that was used on the client?

Reply:
This is called on both servers and clients. Keys need to be worker specific though, as there is no coordinator to prevent clashes.